consensus/src/chained_bft/consensus_types/block.rs

```diff
@@ -218,11 +288,35 @@ where
         validator: &ValidatorVerifier<Ed25519PublicKey>,
     ) -> ::std::result::Result<(), BlockVerificationError> {
         if self.is_genesis_block() {
+            if self.payload != T::default()
+                || self.parent_id != HashValue::zero()
+                || self.round != 0
+                || self.height != 0
+                || self.timestamp_usecs != 0
+            {
+                return Err(BlockVerificationError::InvalidGenesis);
+            }
             return Ok(());
         }
-        validator
-            .verify_signature(self.author(), self.hash(), self.signature())
-            .map_err(|_| BlockVerificationError::SigVerifyError)?;
+        if self.id() != self.hash() {
+            return Err(BlockVerificationError::InvalidBlockId);
+        }
+        if self.quorum_cert().certified_block_id() != self.parent_id() {
+            return Err(BlockVerificationError::QCDoesNotCertifyParent);
+        }
+        if self.quorum_cert().certified_block_round() >= self.round()
+            || self.round() < self.height()
+        {
+            return Err(BlockVerificationError::InvalidBlockRound);
+        }
+        if let BlockSource::Proposal { author, signature } = &self.block_source {
+            validator
+                .verify_signature(*author, self.hash(), signature)
+                .map_err(|_| BlockVerificationError::SigVerifyError)?;
+        } else if self.payload != T::default() {
+            // NIL block must not carry payload
```

Comment:
nit: we can consider having payload as part of BlockSource::RealBlock enum (and rename enum to smt like BlockData). This way NilBlock simply won't have payload.